src/server/engine_factory.rs

```diff
@@ -77,40 +74,31 @@ impl<ER: RaftEngine> KvEngineFactoryBuilder<ER> {
         self
     }
 
-    pub fn compaction_filter_router(mut self, router: RaftRouter<RocksEngine, ER>) -> Self {
-        self.router = Some(router);
+    pub fn compaction_event_sender(
+        mut self,
+        sender: Arc<dyn CompactedEventSender + Send + Sync>,
+    ) -> Self {
+        self.compact_event_sender = Some(sender);
         self
     }
 
-    pub fn build(self) -> KvEngineFactory<ER> {
+    pub fn build(self) -> KvEngineFactory {
         KvEngineFactory {
             inner: Arc::new(self.inner),
-            router: Mutex::new(self.router),
+            compact_event_sender: self.compact_event_sender.clone(),
         }
     }
 }
 
-pub struct KvEngineFactory<ER: RaftEngine> {
+#[derive(Clone)]
+pub struct KvEngineFactory {
     inner: Arc<FactoryInner>,
-    router: Mutex<Option<RaftRouter<RocksEngine, ER>>>,
+    compact_event_sender: Option<Arc<dyn CompactedEventSender + Send + Sync>>,
 }
 
-impl<ER: RaftEngine> Clone for KvEngineFactory<ER> {
-    fn clone(&self) -> Self {
-        Self {
-            inner: self.inner.clone(),
-            router: Mutex::new(self.router.lock().unwrap().clone()),
-        }
-    }
-}
-
-impl<ER: RaftEngine> KvEngineFactory<ER> {
+impl KvEngineFactory {
     pub fn create_raftstore_compaction_listener(&self) -> Option<CompactionListener> {
-        let router = self.router.lock().unwrap();
-        let ch = match &*router {
-            Some(r) => Mutex::new(r.clone()),
-            None => return None,
-        };
+        self.compact_event_sender.as_ref()?;
```

Comment:
What's the point?

Reply:
if compact_event_sender is None, then return immediately.